### PR TITLE
Fix generic signature issue when calling non generic method of closed generic type

### LINKED ISSUE #96469: AssemblyBuilderImpl.DefinePersistedAssembly has trouble with ReadOnlySpan methods
Repro (add this somewhere in the System.Reflection.Emit tests):
```C#
[Fact]
public void Repro()
{
    using (TempFile file = TempFile.Create())
    {
        AssemblyBuilder ab = AssemblySaveTools.PopulateAssemblyBuilderTypeBuilderAndSaveMethod(out TypeBuilder type, out MethodInfo saveMethod);
        MethodBuilder method = type.DefineMethod("M1", MethodAttributes.Public, null, null);

        ILGenerator ilGenerator = method.GetILGenerator();
        LocalBuilder span = ilGenerator.DeclareLocal(typeof(ReadOnlySpan<char>));
        LocalBuilder str = ilGenerator.DeclareLocal(typeof(string));

        ilGenerator.Emit(OpCodes.Ldstr, "hello");
        ilGenerator.Emit(OpCodes.Call, typeof(MemoryExtensions).GetMethod("AsSpan", [typeof(string)])!);
        ilGenerator.Emit(OpCodes.Stloc, span);
        ilGenerator.Emit(OpCodes.Ldloca_S, span);
        ilGenerator.Emit(OpCodes.Ldc_I4_1);
        ilGenerator.Emit(OpCodes.Call, typeof(ReadOnlySpan<char>).GetMethod("Slice", [typeof(int)])!);
        ilGenerator.Emit(OpCodes.Stloc, span);
        ilGenerator.Emit(OpCodes.Ldloca_S, span);
        ilGenerator.Emit(OpCodes.Constrained, typeof(ReadOnlySpan<char>));
        ilGenerator.Emit(OpCodes.Callvirt, typeof(object).GetMethod("ToString"));
        ilGenerator.Emit(OpCodes.Stloc, str);
        ilGenerator.EmitWriteLine(str);
        ilGenerator.Emit(OpCodes.Ret);

        type.CreateType();
        saveMethod.Invoke(ab, [file.Path]);

        Type typeFromDisk = Assembly.LoadFile(file.Path).GetType("MyType");
        typeFromDisk.GetMethod("M1").Invoke(Activator.CreateInstance(typeFromDisk), null);
    }
}
```

That fails with:
```
    Discovering: System.Reflection.Emit.Tests (method display = ClassAndMethod, method display options = None)
    Discovered:  System.Reflection.Emit.Tests (found 632 of 633 test cases)
    Starting:    System.Reflection.Emit.Tests (parallel test collections = on, max threads = 12)
      System.Reflection.Emit.Tests.AssemblySaveMethodBuilderTests.Repro [FAIL]
        System.MissingMethodException : Method not found: 'System.ReadOnlySpan`1<Char> System.ReadOnlySpan`1.Slice(Int32)'.
        Stack Trace:
             at MyType.M1()
             at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
          D:\repos\runtime\src\coreclr\System.Private.CoreLib\src\System\Reflection\MethodBaseInvoker.CoreCLR.cs(36,0): at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args)
          D:\repos\runtime\src\libraries\System.Private.CoreLib\src\System\Reflection\MethodBaseInvoker.cs(57,0): at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)
    Finished:    System.Reflection.Emit.Tests
```

### PR DESCRIPTION
**Pull request by buyaa-n**
The method signature populated incorrectly for non generic method that declared within generic type and uses the generic parameters as return type or parameter.

For example, return type of `ReadonlySpan<T>.Slice(int)` method of an `ReadonlySpan<char>` instance should be closed generic `ReadonlySpan<T>` not `ReadonlySpan<char>`: 

Fixes https://github.com/dotnet/runtime/issues/96469

### CODE REVIEW

File: src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs [hunk 1]
```diff
@@ -2174,5 +2174,39 @@ public void RecursiveSumTest()
...
+                ilGenerator.Emit(OpCodes.Stloc, span);
+                ilGenerator.Emit(OpCodes.Ldloca_S, span);
+                ilGenerator.Emit(OpCodes.Ldc_I4_1);
+                ilGenerator.Emit(OpCodes.Call, typeof(ReadOnlySpan<char>).GetMethod("Slice", [typeof(int)])!);
+                ilGenerator.Emit(OpCodes.Stloc, span);
+                ilGenerator.Emit(OpCodes.Ldloca_S, span);
+                ilGenerator.Emit(OpCodes.Constrained, typeof(ReadOnlySpan<char>));
+                ilGenerator.Emit(OpCodes.Callvirt, typeof(object).GetMethod("ToString"));
+                ilGenerator.Emit(OpCodes.Stloc, str);
+                ilGenerator.EmitWriteLine(str);
```
Comment (stephentoub) [inline src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs:2201]: We shouldn't actually WriteLine unless we're going to validate that stdout gets what we exoect it to

File: src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs [hunk 2]
```diff
@@ -2174,5 +2174,39 @@ public void RecursiveSumTest()
...
+                ilGenerator.Emit(OpCodes.Ret);
+
+                type.CreateType();
+                saveMethod.Invoke(ab, [file.Path]);
+
+                Type typeFromDisk = Assembly.LoadFile(file.Path).GetType("MyType");
```
Comment (stephentoub) [inline src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs:2207]: If we're going to LoadFrom, this test should probably use RemoteExecutor.

File: src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs (hunk 1 above)
Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs:2201]: I can return the string for validation, as don't know how to check what stdout gets.

File: src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs (hunk 2 above)
Comment (buyaa-n) [reply src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs:2207]: Is that for getting the file deleted automatically after the test execution or is there other reason?

Comment (stephentoub) [reply src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs:2207]: In part. The assembly won't be unloaded, so over time such tests will bloat the process.

File: src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Reflection.Emit/tests/PersistableAssemblyBuilder/AssemblySaveILGeneratorTests.cs:2201]: > I can return the string for validation

That's fine.

**Review by stephentoub — APPROVED** [review]

**Pull request merged**